rfcs/20190225-tf-on-demand.md

```diff
@@ -0,0 +1,191 @@
+
+*   [tensorflow-build-s2i ](https://github.com/thoth-station/tensorflow-build-s2i)
+*   [Tensorflow-release-job](https://github.com/thoth-station/tensorflow-release-job)
+*   [tensorflow-serving-build](https://github.com/thoth-station/tensorflow-serving-build)
+
+We propose collaborating with the Red Hat team and leveraging the team's open-source software as much as possible. The server side of the project, in particular, can have concentrated involvement from the team given the backend GKE cluster that already has a lot of functionality built and available. (Please refer to the "Server" block in the diagram above for details on the server-side event flow.) Based on Additional support and functionality will be built around and on top of it to match what TensorFlow currently supports and should support moving forward.
+
+## Design Proposal
+
+### Front-end
+
+**Web UI (download.tensorflow.org)**
+
+We propose a simple web interface that allows users to manually enter necessary system specs and desired build options for requesting a custom TF binary. The UI will be straightforward and easy-to-use. A sample mock up UI is shown below. (Please refer to [PyTorch](https://pytorch.org) download UI as an additional reference.)
+
+![drawing](https://docs.google.com/drawings/d/1Krze2no7zjfqe7nldOm-ArECOGXVVjgtk0VMXCabzkw/export/png)
+
+Once all fields are filled in, the system backend will check if the corresponding binary has already been built and present in cache. If it exists, then it will provide the user a URL to the binary for download. If it does not exist, then it will ask the user for an email address to receive a link for downloading the newly built binary.
+
+Before sending out the binary, the system will first check whether the binary is supported (or unsupported) by going through what is being tested in CI. It will then inform the users accordingly.
+
+**Command Line Tool (TF Downloader Binary)**
+
+We propose a simple binary which will detect and fill out most of the inputs that the Web UI requires. It will then send the request to the backend to build. A sample execution (not final) is shown below:
+
+```
+> python tfdownloader.py
+Downloader will now detect your system:
+	- Detecting CPU…….  Found Intel Core i7-8700K
+	- Detecting GPU……. Found NVIDIA GTX 980
+	- Detecting CUDA…… Found 9.2
+	- Detecting cuDNN….. Found 7.4
+	- Detecting Distribution……. Found Ubuntu 18.04
+
+Requesting TF build with options:
+    CPU options: -mavx2
+    GPU enabled: yes
+    CUDA version: 9.2
+    cuDNN version: 7.4
+    CUDA compute capability: 5.2
+    GCC version: 7.3
+```
+
+
+It is possible that the system this script is run on may be incompatible for running TF. Such incapabilities and what the script is able to do can be categorized into two failure modes:
+
+
+
+*   *Hard Constraint Failures*
+    *   32-bit OS
+    *   GPU too old (Cuda Compute Capability older than 3.0)
```

Comment:
People with old GPUs can still build TF for CPUs, right?

Reply:
Always. And <3.0 was never supported for GPU anyway.